File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/library/items/media/files/nowplaying/menu/NowPlayingFileListItemMenuBuilder.java
package com.lasthopesoftware.bluewater.client.library.items.media.files.nowplaying.menu;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.lasthopesoftware.bluewater.R;
import com.lasthopesoftware.bluewater.client.library.items.media.files.BaseMenuViewHolder;
import com.lasthopesoftware.bluewater.client.library.items.media.files.FilePlayClickListener;
import com.lasthopesoftware.bluewater.client.library.items.media.files.IFile;
import com.lasthopesoftware.bluewater.client.library.items.media.files.ViewFileDetailsClickListener;
import com.lasthopesoftware.bluewater.client.library.items.media.files.menu.AbstractFileListItemNowPlayingHandler;
import com.lasthopesoftware.bluewater.client.library.items.media.files.menu.FileListItemContainer;
import com.lasthopesoftware.bluewater.client.library.items.media.files.menu.FileNameTextViewSetter;
import com.lasthopesoftware.bluewater.client.library.items.media.files.nowplaying.menu.listeners.RemovePlaylistFileClickListener;
import com.lasthopesoftware.bluewater.client.library.items.media.files.playback.service.PlaybackService;
import com.lasthopesoftware.bluewater.client.library.items.media.files.properties.CachedFilePropertiesProvider;
import com.lasthopesoftware.bluewater.client.library.items.menu.AbstractListItemMenuBuilder;
import com.lasthopesoftware.bluewater.client.library.items.menu.LongClickViewAnimatorListener;
import com.lasthopesoftware.bluewater.client.library.items.menu.NotifyOnFlipViewAnimator;
import com.lasthopesoftware.bluewater.shared.view.LazyViewFinder;
import com.lasthopesoftware.bluewater.shared.view.ViewUtils;
import com.vedsoft.futures.runnables.OneParameterAction;

import java.util.List;

/**
 * Created by david on 11/7/15.
 */
public class NowPlayingFileListItemMenuBuilder extends AbstractListItemMenuBuilder<IFile> {

    private static final class ViewHolder extends BaseMenuViewHolder {

        private final LazyViewFinder<ImageButton> removeButtonFinder;

        public ViewHolder(final FileListItemContainer fileListItemContainer, final LazyViewFinder<ImageButton> viewFileDetailsButtonFinder, final LazyViewFinder<ImageButton> playButtonFinder, final LazyViewFinder<ImageButton> removeButtonFinder) {
            super(viewFileDetailsButtonFinder, playButtonFinder);

            this.removeButtonFinder = removeButtonFinder;
            this.fileListItemContainer = fileListItemContainer;
        }

        public final FileListItemContainer fileListItemContainer;
        public AbstractFileListItemNowPlayingHandler fileListItemNowPlayingHandler;
        public CachedFilePropertiesProvider filePropertiesProvider;

	    public final ImageButton getRemoveButton() {
		    return removeButtonFinder.findView();
	    }
    }

    private final List<IFile> files;
    private final int nowPlayingPosition;

    private OneParameterAction<Integer> onPlaylistFileRemovedListener;

    public NowPlayingFileListItemMenuBuilder(final List<IFile> files, final int nowPlayingPosition) {
        this.files = files;
        this.nowPlayingPosition = nowPlayingPosition;
    }

    @Override
    public View getView(final int position, final IFile file, View convertView, ViewGroup parent) {
        if (convertView == null) {
            final FileListItemContainer fileItemMenu = new FileListItemContainer(parent.getContext());
            final NotifyOnFlipViewAnimator viewFlipper = fileItemMenu.getViewAnimator();
            convertView = viewFlipper;

            final LayoutInflater inflater = (LayoutInflater) parent.getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);

            final LinearLayout fileMenu = (LinearLayout)inflater.inflate(R.layout.layout_now_playing_file_item_menu, parent, false);

            viewFlipper.addView(fileMenu);

            viewFlipper.setTag(
		            new ViewHolder(
				            fileItemMenu,
				            new LazyViewFinder<>(fileMenu, R.id.btnViewFileDetails),
				            new LazyViewFinder<>(fileMenu, R.id.btnPlaySong),
				            new LazyViewFinder<>(fileMenu, R.id.btnRemoveFromPlaylist)));
            viewFlipper.setViewChangedListener(getOnViewChangedListener());
        }

        final ViewHolder viewHolder = (ViewHolder)convertView.getTag();

        final FileListItemContainer fileListItem = viewHolder.fileListItemContainer;

        final TextView textView = fileListItem.getTextViewFinder();

        if (viewHolder.filePropertiesProvider != null) viewHolder.filePropertiesProvider.cancel(false);
        viewHolder.filePropertiesProvider = FileNameTextViewSetter.startNew(file, textView);

        textView.setTypeface(null, ViewUtils.getActiveListItemTextViewStyle(position == nowPlayingPosition));

        final int currentPlaylistPosition = PlaybackService.getCurrentPlaylistPosition();
        if (currentPlaylistPosition > -1)
            textView.setTypeface(null, ViewUtils.getActiveListItemTextViewStyle(position == currentPlaylistPosition));

        if (viewHolder.fileListItemNowPlayingHandler != null) viewHolder.fileListItemNowPlayingHandler.release();
        viewHolder.fileListItemNowPlayingHandler = new AbstractFileListItemNowPlayingHandler(fileListItem) {
            @Override
            public void onReceive(Context context, Intent intent) {
                final int playlistPosition = intent.getIntExtra(PlaybackService.PlaylistEvents.PlaylistParameters.playlistPosition, -1);
                textView.setTypeface(null, ViewUtils.getActiveListItemTextViewStyle(position == playlistPosition));
            }
        };

        final NotifyOnFlipViewAnimator viewFlipper = fileListItem.getViewAnimator();
        LongClickViewAnimatorListener.tryFlipToPreviousView(viewFlipper);
        viewHolder.getPlayButton().setOnClickListener(new FilePlayClickListener(viewFlipper, position, files));
        viewHolder.getViewFileDetailsButton().setOnClickListener(new ViewFileDetailsClickListener(viewFlipper, file));
        viewHolder.getRemoveButton().setOnClickListener(new RemovePlaylistFileClickListener(viewFlipper, position, onPlaylistFileRemovedListener));

        return viewFlipper;
    }

    public void setOnPlaylistFileRemovedListener(OneParameterAction<Integer> onPlaylistFileRemovedListener) {
        this.onPlaylistFileRemovedListener = onPlaylistFileRemovedListener;
    }
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/view/nowplaying/menu/NowPlayingFileListItemMenuBuilder.kt
package com.lasthopesoftware.bluewater.client.playback.view.nowplaying.menu

import android.widget.ImageButton
import androidx.recyclerview.widget.RecyclerView
import com.lasthopesoftware.bluewater.R
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.details.ViewFileDetailsClickListener
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.menu.AbstractFileListItemMenuBuilder
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.menu.FileListItemContainer
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.menu.FileListItemNowPlayingRegistrar
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.menu.FileNameTextViewSetter
import com.lasthopesoftware.bluewater.client.browsing.items.menu.LongClickViewAnimatorListener
import com.lasthopesoftware.bluewater.client.playback.file.PositionedFile
import com.lasthopesoftware.bluewater.client.playback.service.broadcasters.PlaylistEvents
import com.lasthopesoftware.bluewater.client.playback.view.nowplaying.menu.listeners.FileSeekToClickListener
import com.lasthopesoftware.bluewater.client.playback.view.nowplaying.menu.listeners.RemovePlaylistFileClickListener
import com.lasthopesoftware.bluewater.client.playback.view.nowplaying.storage.INowPlayingRepository
import com.lasthopesoftware.bluewater.shared.android.view.LazyViewFinder
import com.lasthopesoftware.bluewater.shared.android.view.ViewUtils
import com.lasthopesoftware.bluewater.shared.promises.extensions.LoopedInPromise


class NowPlayingFileListItemMenuBuilder(private val nowPlayingRepository: INowPlayingRepository, private val fileListItemNowPlayingRegistrar: FileListItemNowPlayingRegistrar)
	: AbstractFileListItemMenuBuilder<NowPlayingFileListItemMenuBuilder.ViewHolder>(R.layout.layout_now_playing_file_item_menu) {

	override fun newViewHolder(fileItemMenu: FileListItemContainer) = ViewHolder(fileItemMenu)

	inner class ViewHolder internal constructor(private val fileListItemContainer: FileListItemContainer)
		: RecyclerView.ViewHolder(fileListItemContainer.viewAnimator) {

		private val viewFileDetailsButtonFinder = LazyViewFinder<ImageButton>(itemView, R.id.btnViewFileDetails)
		private val playButtonFinder = LazyViewFinder<ImageButton>(itemView, R.id.btnPlaySong)
		private val removeButtonFinder = LazyViewFinder<ImageButton>(itemView, R.id.btnRemoveFromPlaylist)
		private val fileNameTextViewSetter = FileNameTextViewSetter(fileListItemContainer.findTextView())

		var fileListItemNowPlayingHandler: AutoCloseable? = null

		fun update(positionedFile: PositionedFile) {
			val fileListItem = fileListItemContainer
			val textView = fileListItem.findTextView()

			val serviceFile = positionedFile.serviceFile

			fileNameTextViewSetter.promiseTextViewUpdate(serviceFile)

			val position = positionedFile.playlistPosition

			val viewFlipper = fileListItem.viewAnimator

			nowPlayingRepository
				.nowPlaying
				.eventually(LoopedInPromise.response({ np ->
					textView.setTypeface(null, ViewUtils.getActiveListItemTextViewStyle(position == np.playlistPosition))
					viewFlipper.isSelected = position == np.playlistPosition
				}, textView.context))

			fileListItemNowPlayingHandler?.close()
			fileListItemNowPlayingHandler = fileListItemNowPlayingRegistrar.registerNewHandler(fileListItem) { _, intent ->
				val playlistPosition = intent.getIntExtra(PlaylistEvents.PlaylistParameters.playlistPosition, -1)
				textView.setTypeface(null, ViewUtils.getActiveListItemTextViewStyle(position == playlistPosition))
				viewFlipper.isSelected = position == playlistPosition
			}

			LongClickViewAnimatorListener.tryFlipToPreviousView(viewFlipper)
			playButtonFinder.findView().setOnClickListener(FileSeekToClickListener(viewFlipper, position))
			viewFileDetailsButtonFinder.findView().setOnClickListener(ViewFileDetailsClickListener(viewFlipper, serviceFile))
			removeButtonFinder.findView().setOnClickListener(RemovePlaylistFileClickListener(viewFlipper, position))
		}
	}
}
